pulsar/reader_test.go

```diff
@@ -1023,3 +1024,58 @@ func createPartitionedTopic(topic string, n int) error {
 	}
 	return nil
 }
+
+func TestReaderHasNextFailed(t *testing.T) {
+	client, err := NewClient(ClientOptions{
+		URL: serviceURL,
+	})
+	assert.Nil(t, err)
+	topic := newTopicName()
+	r, err := client.CreateReader(ReaderOptions{
+		Topic:          topic,
+		StartMessageID: EarliestMessageID(),
+	})
+	assert.Nil(t, err)
+	r.(*reader).c.consumers[0].state.Store(consumerClosing)
+	assert.False(t, r.HasNext())
+}
+
+func TestReaderHasNextRetryFailed(t *testing.T) {
+	client, err := NewClient(ClientOptions{
+		URL: serviceURL,
+	})
+	assert.Nil(t, err)
+	topic := newTopicName()
+	r, err := client.CreateReader(ReaderOptions{
+		Topic:          topic,
+		StartMessageID: EarliestMessageID(),
+		// Retry the connection 10 seconds after it's disconnected.
+		BackoffPolicy: newTestBackoffPolicy(10*time.Second, 10*time.Second),
```

Comment:
Maybe we can use a short timeout (3 seconds or 1 second?). 10 seconds are too long for a single test.